docs/architecture/adr-053-go-module-refactoring.md

```diff
@@ -0,0 +1,109 @@
+it has been used in production for years now, not because it isn't production
+quality software, but rather because the API compatibility guarantees required
+by go modules are fairly complex to adhere to with such a large project.
+Up to now, it has generally been deemed more important to be able to break the
+API if needed rather than require all users update all package import paths
+to accommodate breaking changes causing v2, v3, etc. releases. This is in
+addition to the other complexities related to protobuf generated code that will
+be addressed in a separate ADR.
+
+Nevertheless, the desire for semantic versioning has been [strong in the
+community](https://github.com/cosmos/cosmos-sdk/discussions/10162) and the
+single go module release process has made it very hard to
+release small changes to isolated features in a timely manner. Release cycles
+often exceed six months which means small improvements done in a day or
+two get bottle-necked by everything else in the monolithic release cycle.
+
+## Decision
+
+To improve the current situation, the SDK is being refactored into multiple
+go modules within the current repository. There has been a [fair amount of
+debate](https://github.com/cosmos/cosmos-sdk/discussions/10582#discussioncomment-1813377)
+as to how to do this, with some developers arguing for larger vs smaller
+module scopes. There are pros and cons to both approaches (which will be
+discussed below in the [Consequences](#consequences) section), but the
+approach being adopted is the following:
+* a go module should generally be scoped to a specific coherent set of
+functionality (such as math, errors, store, etc.)
+* when code is removed from the core SDK and moved to a new module path, every 
+effort should be made to avoid API breaking changes in the existing code using
+aliases and wrapper types (as done in https://github.com/cosmos/cosmos-sdk/pull/10779
+and https://github.com/cosmos/cosmos-sdk/pull/11788)
+* new go modules should be moved to a standalone domain (`cosmossdk.io`) before
+being tagged as `v1.0.0` to accommodate the possibility that they may be
+better served by a standalone repository in the future
+* all go modules should follow the guidelines in https://go.dev/blog/module-compatibility
+before `v1.0.0` is tagged and should make use of `internal` packages to limit
+the exposed API surface
+* the new go module's API may deviate from the existing code where there are
+clear improvements to be made or to remove legacy dependencies (for instance on
+amino or gogo proto), as long the old package attempts
+to avoid API breakage with aliases and wrappers
+* care should be taken when simply trying to turn an existing package into a
+new go module: https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository.
+In general, it seems safer to just create a new module path (appending v2, v3, etc.
+if necessary), rather than trying to make an old package a new module.
```

Comment:
This sentence is a bit confusing to me. It seems for `errors`, we did create a new module from an old package.

Reply:
Another thing that's not clear for me is: when bumping a major version, do we:
- rename in go.mod `cosmossdk.io/foo` -> `cosmossdk.io/foo/v2`, and tag `foo/v2.0.0`
- or no rename in go.mod, simply tag `foo/v2.0.0`

Reply:
My understanding is that go.mod `module` directive has to be updated, but moving code to subdirectory is optional:

> If the module is released at major version 2 or higher, the module path must end with a [major version suffix](https://go.dev/ref/mod#major-version-suffixes) like /v2. This may or may not be part of the subdirectory name. For example, the module with path golang.org/x/repo/sub/v2 could be in the /sub or /sub/v2 subdirectory of the repository golang.org/x/repo.

ref: https://go.dev/ref/mod